### PR TITLE
Added tests for WKO Retry Improvements: Add a test in operator usability for failureRetryLimitMinutes

### PR DESCRIPTION
**Pull request by hzhao-github**
new PR https://github.com/oracle/weblogic-kubernetes-operator/pull/3523

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java [hunk 1]
```diff
@@ -283,8 +307,8 @@ void testRetryOccursAsExpectedAndThrowWarning() {
         logger, "warningMsgRegex is found in domain status message");
 
     // verify that WARNING and warningMsgRegex message found in Operator log
-    testUntil(() -> findStringInOperatorLog(opNamespace, ".*WARNING" + warningMsgRegex),
-        logger, "warningMsgRegex is found in Operator log");
+    testUntil(() -> checkPodLogContainsRegex(warningMsgRegex, operatorPodName, opNamespace),
+        logger, "{0}} is found in Operator log", warningMsgRegex);
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java:311]: {0}} -> {0}

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java [hunk 2]
```diff
@@ -313,26 +337,100 @@ void testRetryOccursAsExpectedAndThrowWarning() {
...
     // verify that the cluster is still up and running
     verifyDomainExistsAndServerStarted(replicaMaxCount);
   }
 
+  /**
+   * Create a domain-in-image domain before the secret for admin credentials is created.
+   * Verify that retry stops after failureRetryLimitMinutes expired.
+   */
+  @Test
+  @DisplayName("Verify that retry stops after the issue is fixed and the domain starts successfully.")
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java:351]: The java doc and the @DisplayName content is not consistent.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java [hunk 3]
```diff
@@ -313,26 +337,100 @@ void testRetryOccursAsExpectedAndThrowWarning() {
...
+        logger, "{0} is found in Operator log", createDomainFailedMsgRegex);
+
+    // verify that SEVERE and createDomainFailedMsgRegex message found in Operator log
+    testUntil(() -> checkInUncompletedIntroPodLogContainsRegex(createDomainFailedMsgRegex,
+        domainUid, domainNamespace),
+        logger, "{0} is found in introspector log", createDomainFailedMsgRegex);
+
+    Callable<Boolean> configMapExist = assertDoesNotThrow(() -> configMapExist(domainNamespace, badModelFileCm));
+
+    if (configMapExist.call().booleanValue()) {
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java:414]: You can use 
if (configMapExist(domainNamespace, badModelFileCm)) directly.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DomainUtils.java [hunk 4]
```diff
@@ -991,6 +992,170 @@ public static DomainResource createDomainInImageAndVerify(String domainUid,
...
+                        .channelName("default")
+                        .nodePort(getNextFreePort()))))
+            .configuration(new Configuration()
+                .model(new Model()
+                    .domainType(WLS_DOMAIN_TYPE)
+                    .configMap(configmapName))
+            .introspectorJobActiveDeadlineSeconds(300L)));
+
+    // create cluster resource for the domain
+    if (!Cluster.doesClusterExist(clusterName, CLUSTER_VERSION, domainNamespace)) {
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DomainUtils.java:1149]: You can use static import Cluster.doesClusterExist() here.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/PodUtils.java [hunk 5]
```diff
@@ -454,4 +458,55 @@ public static boolean checkPodLogContains(String matchStr, String podName, Strin
...
+   * @return true if regex found, false otherwise.
+   */
+  public static boolean checkInUncompletedIntroPodLogContainsRegex(String regex, String domainUid, String namespace) {
+    // get introspector pod name
+    String introspectorPodName = assertDoesNotThrow(()
+        -> getIntrospectorPodName(domainUid, namespace), "Getting introspector pod name failed");
+    getLogger().info("introspector pod name is: {0}", introspectorPodName);
+
+    // get the domain status message
+    StringBuffer getPodLogCmd = new StringBuffer("kubectl logs pod/");
```
Comment (xiancao) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/PodUtils.java:495]: You can use TestActions.getPodLog() directly here.

**Review by xiancao — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java (hunk 3 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java:414]: Sorry I can't. configMapExist.call().booleanValue() returns Callable<Boolean> but if requires boolean

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/PodUtils.java (hunk 5 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/PodUtils.java:495]: I first tried TestActions.getPodLog() but it didn't work for this case, this method gets logs from a completed pod but this case needs logs from an incomplete pod :(

**Comment by hzhao-github** [comment]
please refer to PR https://github.com/oracle/weblogic-kubernetes-operator/pull/3523

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java (hunk 1 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovements.java:311]: changed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DomainUtils.java (hunk 4 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DomainUtils.java:1149]: changed

**Pull request closed without merging**